internal/scheduler/nodedb.go

```diff
@@ -0,0 +1,332 @@
+		matches, reason, err := nodeType.PodRequirementsMet(req)
+		if err != nil {
+			return nil, nil, err
+		}
+		if matches {
+			selectedNodeTypes = append(selectedNodeTypes, nodeType)
+		} else if reason != nil {
+			numNodeTypesExcludedByReason[reason.String()] += 1
+		} else {
+			numNodeTypesExcludedByReason["unknown reason"] += 1
+		}
+	}
+	return selectedNodeTypes, numNodeTypesExcludedByReason, nil
+}
+
+func (nodeDb *NodeDb) dominantResource(req *schedulerobjects.PodRequirements) string {
+	dominantResourceType := ""
+	dominantResourceFraction := 0.0
+	for t, q := range req.ResourceRequirements.Requests {
+		// Skip any resource types that are not indexed.
+		if _, ok := nodeDb.indexedResources[string(t)]; !ok {
+			continue
+		}
+		available, ok := nodeDb.totalResources[string(t)]
+		if !ok {
+			return string(t)
+		}
+		f := q.AsApproximateFloat64() / available.AsApproximateFloat64()
+		if f >= dominantResourceFraction {
+			dominantResourceType = string(t)
+			dominantResourceFraction = f
+		}
+	}
+	return dominantResourceType
+}
+
+// MarkJobRunning notifies the node db that this job is now running.
+// When the nodes were bound to the job, resources on those nodes were marked as assigned in the node db.
+// When the job is running, those resources are accounted for by the executor,
+// and should no longer be marked as assigned in the node db.
+//
+// TODO: This only clears AssignedByNode once there are no in-flight jobs for that node.
+// We could improve it to clear AssignedByNode on a per-job basis.
+func (nodeDb *NodeDb) MarkJobRunning(jobId uuid.UUID) {
+	for nodeId := range nodeDb.NodesByJob[jobId] {
+		delete(nodeDb.JobsByNode[nodeId], jobId)
+		if len(nodeDb.JobsByNode[nodeId]) == 0 {
+			delete(nodeDb.AssignedByNode, nodeId)
+		}
+	}
+	delete(nodeDb.NodesByJob, jobId)
+}
+
+func NewNodeDb(priorities []int32, resourceTypes []string) (*NodeDb, error) {
+	db, err := memdb.NewMemDB(nodeDbSchema(priorities, resourceTypes))
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	priorities = []int32(priorities)
+	slices.Sort(priorities)
+	totalResources := make(map[string]*resource.Quantity)
+	indexedResources := make(map[string]interface{})
+	for _, resourceType := range resourceTypes {
+		q := resource.MustParse("0")
+		totalResources[resourceType] = &q
+		indexedResources[resourceType] = true
+	}
+	return &NodeDb{
+		priorities:       priorities,
+		indexedResources: indexedResources,
+		NodeTypes:        make(map[string]*schedulerobjects.NodeType),
+		totalResources:   totalResources,
+		Db:               db,
+		NodesByJob:       make(map[uuid.UUID]map[string]interface{}),
+		JobsByNode:       make(map[string]map[uuid.UUID]interface{}),
+		AssignedByNode:   make(map[string]schedulerobjects.AssignedByPriorityAndResourceType),
+	}, nil
+}
+
+// Upsert will update the node db with the given nodes.
+func (nodeDb *NodeDb) Upsert(nodes []*schedulerobjects.Node) error {
+	txn := nodeDb.Db.Txn(true)
+	defer txn.Abort()
```

Comment:
This is a bit weird- is it idiomatic to call commit() then abort() on success?

Reply:
It's the way they recommend in their docs. `Abort` is a no-op if read-only or already committed.